### PR TITLE
Feature/co 1466 implement unit tests

### PR DESCRIPTION
**Pull request by kalcicd**

### CODE REVIEW

**Comment by kalcicd** [comment]
Do you think that any more tests are required for the endpoint functions? If not, I'll remove the endpoint tests from the forEach loops. Also, is there merit to writing a test for the valueOperations helper?

File: src/tests/unit/directory-dao.test.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+const proxyDao = (option) => {
+  const proxyquireObject = {
+    util: {},
+
+    './connection': {
+      getClient: sinon.stub().returns({ search: {} }),
+    },
+  };
+
+  if (option === 'getDirectories') {
```
Comment (Ruefa) [inline src/tests/unit/directory-dao.test.js]: There's a lot of repeated code in here. Here's how I would reduce it.
```
proxyquireObject.util.promisify = () => sinon.stub().resolves({
      on: (type, listener) => {
        if (type === option) {
          listener(option === 'searchEntry' ? { object: {} } : undefined);
        }
      },
    });
```
You can then pass in 'done' and 'searchEntry' for option and you won't need the if statement at all.

File: src/tests/unit/directory-dao.test.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    });
+  }
+  return proxyquire('api/v2/db/ldap/directory-dao', proxyquireObject);
+};
+
+describe('Test directory-dao', () => {
+  afterEach(() => sinon.restore());
+
+  describe('Test getDirectory', () => {
+    const testCaseList = [singleResult];
```
Comment (Ruefa) [inline src/tests/unit/directory-dao.test.js]: Don't need to loop if there is only one test case.

File: src/tests/unit/directory-dao.test.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        serializeDirectoryStub.returns(testCase);
+        const result = proxyDao('getDirectory').getDirectory('fakeId');
+        return result.should.eventually.be.fulfilled.and.deep.equal(expectedResult);
+      });
+    });
+  });
+
+
+  describe('Test getDirectories', () => {
+    const testCaseList = [multiResult];
```
Comment (Ruefa) [inline src/tests/unit/directory-dao.test.js]: Same as above. one test case doesn't need a loop.

File: src/tests/unit/directory-dao.test.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    _.forEach(testCaseList, ({ testCase, expectedResult, description }) => {
+      it(`getDirectories should be fulfilled with ${description}`, () => {
+        const serializeDirectoriesStub = sinon.stub(directorySerializer, 'serializeDirectories');
+        serializeDirectoriesStub.returns(testCase);
+        const result = proxyDao('getDirectories').getDirectories({ firstName: 'fakeName' });
+        return result.should.eventually.be.fulfilled.and.deep.equal(expectedResult);
+      });
+    });
```
Comment (Ruefa) [inline src/tests/unit/directory-dao.test.js]: These two cases are near identical. I would make a helper function that takes the differences as parameters. Having the serializer stub return `testCase` which is the same as `expectedResult` pretty much nullifies your tests. Try setting it to `returnsArg(0)`, I don't think that will break your tests. Then you can just always stub both serializers in the helper function and it won't need to be an additional parameter. Checkout the helper functions I wrote in [onbase unit tests](https://github.com/osu-mist/onbase-api/blob/master/src/tests/unit/onbase-dao.test.js) for inspiration.

File: src/tests/unit/directory-dao.test.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import chai from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import _ from 'lodash';
+import sinon from 'sinon';
+import proxyquire from 'proxyquire';
```
Comment (Ruefa) [inline src/tests/unit/directory-dao.test.js:4]: imports should be in alphabetical order based on the package/file name. If going by file names (the second section of imports, line 7-8) then you include the directory prefix i.e. the whole `'api/v2/serializers/directory-serializer'` would be included in the alphabetization. `proxyquire` is the only one out of the place here.

File: src/tests/unit/test-util.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
+
```
Comment (Ruefa) [inline src/tests/unit/test-util.js]: Remove this extra line

File: src/tests/unit/test-util.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+import sinon from 'sinon';
+import { expect } from 'chai';
+import _ from 'lodash';
+import * as conn from 'api/v2/db/ldap/connection';
+import { openapi } from 'utils/load-openapi';
+import { fakeBaseUrl } from './mock-data';
+// import { clear } from 'winston';
```
Comment (Ruefa) [inline src/tests/unit/test-util.js]: remove comment

File: src/tests/unit/test-util.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
+
+import sinon from 'sinon';
```
Comment (Ruefa) [inline src/tests/unit/test-util.js]: Alphabetize imports. See above comment.

File: src/tests/unit/test-util.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import { expect } from 'chai';
+import _ from 'lodash';
+import * as conn from 'api/v2/db/ldap/connection';
```
Comment (Ruefa) [inline src/tests/unit/test-util.js]: local imports should be separate from package imports. Add a line here between conn and lodash similar to directory-dao-test.js.

File: src/tests/unit/test-util.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+import { openapi } from 'utils/load-openapi';
+import { fakeBaseUrl } from './mock-data';
+// import { clear } from 'winston';
+
+
+/**
+ * @summary Create the stub for client object that is useful for controlled testing environment.
+ * @function
+ * @param {object} testCase The testcase that becomes the return value of client.promiseSearch
```
Comment (Ruefa) [inline src/tests/unit/test-util.js]: Looks like this param is no longer included.

File: src/tests/unit/test-util.js (hunk 6 above)
Comment (Ruefa) [inline src/tests/unit/test-util.js]: This file isn't used anywhere. I'm assuming they are for the serializer tests which are also missing.

**Review by Ruefa — CHANGES_REQUESTED** [review]
I got an error when trying to run without a config file. I did the same thing while writing the onbase unit tests. See how I stubbed config [here](https://github.com/osu-mist/onbase-api/blob/master/src/tests/unit/onbase-dao.test.js). It also looks like the serializer tests are missing.

File: src/api/v2/serializers/directory-serializer.js [hunk 11]
```diff
@@ -92,6 +92,7 @@ const serializeDirectories = (rawDirectories, query) => {
     });
   });
 
+
```
Comment (Ruefa) [inline src/api/v2/serializers/directory-serializer.js]: remove extra line

File: src/tests/unit/directory-dao.test.js [hunk 12]
```diff
@@ -1,74 +1,66 @@
...
+      }),
+    },
     './connection': {
       getClient: sinon.stub().returns({ search: {} }),
     },
-  };
+    '../../serializers/directory-serializer': {
+      serializeDirectory: anonStub,
+      serializeDirectories: anonStub,
+    },
```
Comment (Ruefa) [inline src/tests/unit/directory-dao.test.js]: removed extra line

File: src/tests/unit/directory-serializer.test.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import chai from 'chai';
```
Comment (Ruefa) [inline src/tests/unit/directory-serializer.test.js]: You should take a look at some of the test helpers [here](https://github.com/osu-mist/game-curator-api/blob/master/tests/unit/test-helpers.js#L45). Almost any change to the swagger document will break your tests. You should be able to copy over those helpers without making too many changes.

**Review by Ruefa — CHANGES_REQUESTED** [review]

File: src/tests/unit/test-utils.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+  expect(serializedResource).to.deep.equal(schema);
+};
+
+/**
+ * Validate multiple serialized resources.
+ *
+ * @param {object} serializedResources serialized resource to be tested.
+ * @param {object} rawResources Raw resources to be used in test.
+ * @param {*} resourceType The type of the resource.
+ * @param {*} resourceKey The id of the resource.
```
Comment (Ruefa) [inline src/tests/unit/test-utils.js]: These params should be strings

File: src/tests/unit/test-utils.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
+import { expect } from 'chai';
+import _ from 'lodash';
+
+import { openapi } from 'utils/load-openapi';
+import { fakeBaseUrl } from './mock-data';
+import { valueOperations, ldapKeyToResourceKey } from '../../api/v2/serializers/directory-serializer';
+
+
```
Comment (Ruefa) [inline src/tests/unit/test-utils.js]: remove extra line

**Review by Ruefa — APPROVED** [review]
Fix the 2 small issues then you can merge.

**Pull request merged**